### PR TITLE
[narwhal] Use CertificateV2 to speed up narwhal catchup

### PR DESCRIPTION
**Pull request by arun-koshy**
## Description 

[PR#13777](https://github.com/MystenLabs/sui/pull/13777) introduces new `SignatureVerificationState` that can now be used in certificate fetching to only verify the tip of certificate chains and ensure that the verification state is reflected in storage. Doing so should save us time in signature verification for a node that is trying to catchup.

## Test Plan 

Added unit tests. Catchup tests in private-testnet.

#### NW Catchup Rate @ [200 TPS w/ 2 hours of downtime](https://metrics.sui.io/d/ORCQSHfVk/nw-catchup-dashboard?var-Environment=8Xt1pVoVk&var-network=private-testnet&var-validator=ams-ptn-val-00&var-validator=ams-ptn-val-09&orgId=1&from=1696385905867&to=1696388154869)
Screenshot 2023-10-10 at 4 25 53 PM

#### NW Catchup Rate @ [5K TPS  w/ 1 hour of downtime](https://metrics.sui.io/d/ORCQSHfVk/nw-catchup-dashboard?var-Environment=8Xt1pVoVk&var-network=private-testnet&var-validator=ams-ptn-val-00&var-validator=ams-ptn-val-02&var-validator=ams-ptn-val-09&var-validator=del-ptn-val-08&var-validator=sjc-ptn2-val-00&var-validator=ams-ptn-val-03&orgId=1&from=1696643114770&to=1696651260983)
Screenshot 2023-10-10 at 4 26 24 PM

#### Narwhal catchup only hits full potential after state sync completes
Screenshot 2023-10-10 at 4 31 21 PM

#### Execution bottleneck at high TPS preventing higher NW catchup rate
Screenshot 2023-10-10 at 4 33 39 PM

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [X] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

Protocol upgrade to version 28 which will enable the use of CertificateV2 in narwhal that will be used to speed up processing of certificates during certificate fetching/catchup.

### CODE REVIEW

File: narwhal/primary/src/primary.rs [hunk 1]
```diff
@@ -962,6 +979,23 @@ impl PrimaryToPrimary for PrimaryReceiverHandler {
                 .map_err(|e| anemo::rpc::Status::from_error(Box::new(e)))?
             {
                 Some(cert) => {
+                    if self.protocol_config.narwhal_certificate_v2() {
```
Comment (mwtian) [inline narwhal/primary/src/primary.rs:982]: Skipping serving a certificate may make subsequent certificates miss parents. We need to ensure all certificates served as fetch result can be verified directly or indirectly. And the optimization is to skip direct verifications when possible on the fetcher side.

File: narwhal/primary/src/certificate_fetcher.rs [hunk 2]
```diff
@@ -474,7 +533,108 @@ async fn process_certificates_helper(
...
+    }
+
+    let all_certificates_count = all_certificates.len() as u64;
+
+    // Identify leaf certs and preemptively set the parent certificates
+    // as verified indirectly. This is safe because any leaf certs that
+    // fail verification will cancel processing for all fetched certs.
+    let mut leaf_certs = Vec::new();
+    for (idx, c) in all_certificates.iter_mut().enumerate() {
+        if !all_parents.contains(&c.digest()) {
```
Comment (mwtian) [inline narwhal/primary/src/certificate_fetcher.rs:572]: This is not enough to make sure all certificates have parents available in store or among other fetched certificates. e.g. a fetch certificate can exist in all_parents, but itself is missing some parents.

Comment (arun-koshy) [reply narwhal/primary/src/certificate_fetcher.rs:572]: But that is all checked in `try_accept_fetched_certificate` so it is okay, right? Only thing we are skipping is the signature verification.

File: narwhal/primary/src/primary.rs (hunk 1 above)
Comment (arun-koshy) [reply narwhal/primary/src/primary.rs:982]: As discussed switching to serve ALL certificates in store and leave it up to the fetcher to verify.

**Review by mwtian — APPROVED** [review]

**Pull request merged**